pkg/app/zipper/tldcache.go

```diff
@@ -34,12 +45,68 @@ func (app *App) probeTopLevelDomains(ms *PrometheusMetrics) {
 	}
 }
 
+type tldPrefix struct {
+	prefix  string
+	nsCount int
+}
+
+func sortedByNsCount(prefixes []string) []string {
+	countedPrefixes := make([]tldPrefix, len(prefixes))
+	for i, prefix := range prefixes {
+		if prefix == "" {
+			countedPrefixes[i] = tldPrefix{
+				prefix:  prefix,
+				nsCount: 0,
+			}
+		}
+		countedPrefixes[i] = tldPrefix{
+			prefix:  prefix,
+			nsCount: strings.Count(prefix, ".") + 1,
+		}
+	}
+	sort.Slice(countedPrefixes, func(i, j int) bool {
+		return countedPrefixes[i].nsCount < countedPrefixes[j].nsCount
+	})
+	sortedPrefixes := make([]string, len(prefixes))
+	for i := range countedPrefixes {
+		sortedPrefixes[i] = countedPrefixes[i].prefix
+	}
+	return sortedPrefixes
+}
+
+func getBackendsForPrefix(prefix string, backends []backend.Backend, tldCache map[string][]*backend.Backend) []*backend.Backend {
+	var bs []*backend.Backend
+	for {
+		if prefixBackends, ok := tldCache[prefix]; ok {
+			bs = prefixBackends
+			break
+		}
+		lastDotIndex := strings.LastIndex(prefix, ".")
+		if lastDotIndex == -1 {
+			for i := range backends {
+				bs = append(bs, &backends[i])
+			}
+			break
+		}
+		prefix = trimPrefix(prefix[:lastDotIndex])
```

Comment:
imho we shouldn't be modifying input arg, but have a separate variable for mutation so that it's easier to see lifecycle of the variable